sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/dsl/package.scala

```diff
@@ -271,8 +271,9 @@ package object dsl {
       override def expr: Expression = Literal(s)
       def attr: UnresolvedAttribute = analysis.UnresolvedAttribute(s)
     }
-    implicit class DslAttr(attr: UnresolvedAttribute) extends ImplicitAttribute {
-      def s: String = attr.name
+    implicit class DslAttr(a: UnresolvedAttribute) extends ImplicitAttribute {
+      def s: String = a.name
+      override def attr: UnresolvedAttribute = a
```

Comment:
let's fix this completely so that people won't step into this pitfall again.

first problem:
```
implicit class DslAttr(a: UnresolvedAttribute) ... {
  def s: String = a.name
}
abstract class ImplicitAttribute ... {
  def attr: UnresolvedAttribute = analysis.UnresolvedAttribute(s)
}
```
I think the real problem is `UnresolvedAttribute.name` can't be parsed back so this roundtrip is broken. We should update `def s` to
```
def s: String = a.sql
```

The second problem:
```
implicit class DslAttr(a: UnresolvedAttribute) ... {
  def s: String = a.name
}
abstract class ImplicitAttribute ... {
  def int: AttributeReference = AttributeReference(s, IntegerType, nullable = true)()
}
```
The `s` is the qualified column name, and we should handle it properly when creating `AttributeReference`
```
def int: AttributeReference = AttributeReference(attr.nameParts.head, IntegerType, nullable = true)(qualifier = attr.nameParts.init)
```

Reply:
overriding `attr` in `DslAttr` is just an optimization to avoid the roundtrip, but it does not fix the problem, just hides it.

Reply:
Yeah, this seems worth fixing.

Some tests have been relying on multi-part attributes implicitly becoming single-part references. And we have to update those tests now to correct the behavior for multi-part attributes. I don't know if you consider that a wider break. It seems the DSL is solely used for testing, but technically these are public classes.